paddle/framework/recurrent_network_op.cc

```diff
@@ -0,0 +1,224 @@
+  }
+  // prepare inlinks
+  PADDLE_ENFORCE(inlinks_.empty(), "RecurrentOp duplicate inited");
+  for (auto id : GetAttr<std::vector<int>>("real_input")) {
+    inlinks_.push_back(inputs_[id]);
+  }
+
+  name_ = op_desc.name();
+  net_name_ = inputs_.at(GetAttr<int>("step_net"));
+  step_scopes_name_ = inputs_.at(GetAttr<int>("step_scopes"));
+
+  // set memories
+  auto memories = GetAttr<std::vector<std::string>>("memories");
+  auto pre_memories = GetAttr<std::vector<std::string>>("pre_memories");
+  PADDLE_ENFORCE(memories.size() == pre_memories.size(),
+                 "The size of memories and pre_memories doesn't match: %d,%d.",
+                 memories.size(), pre_memories.size());
+  std::vector<std::string> boot_memories;
+  for (auto id : GetAttr<std::vector<int>>("boot_memories")) {
+    boot_memories.push_back(inputs_[id]);
+  }
+  PADDLE_ENFORCE(memories.size() == boot_memories.size(),
+                 "the size of memories and boot_memories doesn't match: %d,%d",
+                 memories.size(), boot_memories.size());
+  for (size_t i = 0; i < memories.size(); ++i) {
+    MemoryAttr mem_attr;
+    mem_attr.var = memories[i];
+    mem_attr.pre_var = pre_memories[i];
+    mem_attr.boot_var = boot_memories[i];
+    memory_attrs_.push_back(mem_attr);
+    LOG(INFO) << "set memorys:\t"
+              << "memory:" << mem_attr.var << "\tboot:" << mem_attr.boot_var;
+  }
+}
+
+void RecurrentOp::CreateScopes(ScopePtr scope) const {
+  LOG(INFO) << "create scopes";
+  auto dims = Input(scope, 0)->GetMutable<Tensor>()->dims();
+  size_t seq_len = dims[0];
+  Variable* scopes_var = scope->GetVariable(step_scopes_name_);
+  auto step_scopes = scopes_var->GetMutable<std::vector<ScopePtr>>();
+  // TODO Only two scopes are needed for inference, this case will be
+  // supported later.
+  if (seq_len > step_scopes->size()) {
+    for (size_t i = step_scopes->size(); i < seq_len; ++i) {
+      step_scopes->push_back(std::make_shared<Scope>(scope));
+    }
+  }
+}
+
+void RecurrentOp::CreateStepNet(ScopePtr scope) const {
+  Variable* var = scope->CreateVariable(net_name_);
+  auto step_net = GetAttr<std::string>("step_net");
+  // get the step net proto from the string.
+  // PADDLE_ENFORCE(
+  //   google::protobuf::TextFormat::ParseFromString(step_net,
+  //   &step_net_desc_));
+  // var->Reset<PlainNet>(new PlainNet(step_net_desc_));
+  // this is a fake net, it will be rewrite after the network has been merged.
+  NetDesc desc;
+  desc.name_ = "rnn_step_net";
+  var->Reset<PlainNet>(new PlainNet(desc));
+  // TODO add op descs
+}
+
+void RecurrentOp::SegmentInputs(ScopePtr scope) const {
+  PADDLE_ENFORCE(!inlinks_.empty(), "no real inputs are provided.");
+  Variable* scopes_var = scope->GetVariable(step_scopes_name_);
+  auto& step_scopes = *scopes_var->GetMutable<std::vector<Scope*>>();
+
+  auto dims = Input(scope, 0)->GetMutable<Tensor>()->dims();
+  int seq_len = dims[0];
+  int batch_size = dims[1];
```

Comment:
ddims is something like `(sent_len, batchsize, ...)`

sometimes input tensor has more dimensions than 3, for example `(sent_len, batchsize, n, m)`, for each step, input is a tensor with shape `(batchsize, n, m)` .

`input_dim = input_dims[2]` -> something like `input_shape = input_dims[2:]` ？ @luotao1